### PR TITLE
Added Semaphore

### PR DESCRIPTION
**Pull request by sourabhpoddar404**
Implemented a solution to add the semaphore but there is an issue with threads.

### CODE REVIEW

File: src/main/java/org/hobbit/core/components/AbstractTaskGenerator.java [hunk 1]
```diff
@@ -61,6 +62,8 @@
      * Mutex used to wait for the start signal after the component has been started
      * and initialized.
      */
+    final private Semaphore currentlyProcessedMessages = new Semaphore(0);
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractTaskGenerator.java]: Inserting the new attribute here is not a good idea – simply because it has the javadoc comment of the `startTaskGenMutex` attribute. It should have its own javadoc comment and all other attributes should of course keep their javadoc comments.

File: src/main/java/org/hobbit/core/components/AbstractTaskGenerator.java [hunk 2]
```diff
@@ -140,9 +143,11 @@ public void handleData(byte[] data) {
 
     @Override
     public void run() throws Exception {
+
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractTaskGenerator.java]: Please avoid the addition of empty lines.

File: src/main/java/org/hobbit/core/components/AbstractTaskGenerator.java [hunk 3]
```diff
@@ -155,7 +160,9 @@ public void run() throws Exception {
     @Override
     public void receiveGeneratedData(byte[] data) {
         try {
+            currentlyProcessedMessages.acquire();
             generateTask(data);
+            currentlyProcessedMessages.release(1);
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractTaskGenerator.java]: Tthe release here is not safe. Please add an additional `try-finally` block to ensure that the permit is released even if `generateTask(data)` throws an exception.

File: src/test/java/org/hobbit/core/components/BenchmarkControllerTest.java [hunk 4]
```diff
@@ -171,6 +171,7 @@ public void init() throws Exception {
     protected void executeBenchmark() throws Exception {
         // give the start signals
         sendToCmdQueue(Commands.TASK_GENERATOR_START_SIGNAL);
+
```
Comment (MichaelRoeder) [inline src/test/java/org/hobbit/core/components/BenchmarkControllerTest.java]: This line is not necessary and it is the only change in this class. So please remove it to reduce the number of unnecessary changes.

File: src/test/java/org/hobbit/core/components/TaskGeneratorTest.java [hunk 5]
```diff
@@ -67,7 +67,7 @@
         // processing (max 100)
         testConfigs.add(new Object[] { 1, 10000, 100, 0 });
         // We use two data generators without parallel message processing
-        testConfigs.add(new Object[] { 2, 10000, 1, 0 });
+        testConfigs.add(new Object[] { 1, 10000, 1, 0 });
```
Comment (MichaelRoeder) [inline src/test/java/org/hobbit/core/components/TaskGeneratorTest.java]: Please revert this change as it changes the test configuration in a way that it does not fit to the description.

File: src/test/java/org/hobbit/core/components/TaskGeneratorTest.java [hunk 6]
```diff
@@ -109,7 +109,7 @@ public void close() throws IOException {
         super.close();
     }
 
-    @Test(timeout = 60000)
+    @Test(timeout = 600000)
```
Comment (MichaelRoeder) [inline src/test/java/org/hobbit/core/components/TaskGeneratorTest.java]: Why is the timeout increased? Why is it increased ten times?

**Review by MichaelRoeder — CHANGES_REQUESTED** [review]
The main idea is as discussed. However, some changes are not clear to me and the handling of the semaphore has to be improved.

**Comment by Yamini19** [comment]
Changes done as suggested

File: src/main/java/org/hobbit/core/components/AbstractTaskGenerator.java [hunk 7]
```diff
@@ -17,6 +17,7 @@
 package org.hobbit.core.components;
 
 import java.io.IOException;
+import java.sql.SQLOutput;
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractTaskGenerator.java]: Why is this import necessary?

**Review by MichaelRoeder — COMMENTED** [review]
Please remove the unused import. Apart from that, the solution looks good.

**Review by MichaelRoeder — APPROVED** [review]
LGTM
